lib/framework/procbot.ts

```diff
@@ -234,4 +235,43 @@ export class ProcBot {
 
 		return service;
 	}
+
+	/**
+	 * Create a new ServiceListener or ServiceEmitter.
+	 * Uses the passed name or handle as the indentifier for future creation/lookups.
+	 * Should an instance with the handle (or name) already exist then this will be passed back.
+	 *
+	 * @param type  Type of Service to create.
+	 * @param name  Name of the service to create.
+	 * @param data  ServiceBase extended constructor, giving an optional handle and constructor data.
+	 * @return      A ServiceListener, ServiceEmitter or void value.
+	 */
+	private addService(type: ServiceType, name: string, data?: ServiceConstructor):
+	ServiceListener | ServiceEmitter | void {
+		// If there's no handle in the constructor data, the handle is the name.
+		let handle = data ? data.handle : name;
+		if (!handle) {
+			handle = name;
+		}
+
+		// Attempt to retrieve the named service.
+		const service = this.getService(name);
+		const serviceMap = (type === ServiceType.Listener) ? this.listeners.get(handle) : this.emitters.get(handle);
```

Comment:
There's an issue with this though. We check types before we create Service instances. Which means we'd need an implementation in the `ProcBot` class to test. However, there's a couple of other tests in the Service themselves, which won't be touching the `ProcBot` class, so we'd need to define them in the ServiceListener/ServiceEmitter themselves. So we'd end up with two places we'd need to change them anyway.

I want to have a think about this; so, I'm not going to do it now. But it's a very valid suggestion.